### PR TITLE
HHH-13105 - Add registries for user-defined Types as currently a custom Type is instantiated and configured 9 times

### PR DESCRIPTION
**Pull request by vladmihalcea**
https://hibernate.atlassian.net/browse/HHH-13105

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/type/TypeFactory.java [hunk 1]
```diff
@@ -52,6 +55,12 @@
 	private final TypeConfiguration typeConfiguration;
 	private final TypeScope typeScope;
 
+	private final Map<TypeKey, Type> typeRegistry = new HashMap<>();
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/type/TypeFactory.java]: This was my concern about boot versus run time.  These Maps live on during run time even though most of them will never be used beyond boot time.  But this class is already deprecated in anticipation of 6 so it's probably not overly relevant

Comment (vladmihalcea) [reply hibernate-core/src/main/java/org/hibernate/type/TypeFactory.java]: That's true. In fact, the Type is associated with the property in the entity Persister, and that's how it's used after boot.

I inspired myself from the basic type registry. If you think this is fine, I'll push it in order to speed up the bootstrapping a little bit.

Comment (vladmihalcea) [reply hibernate-core/src/main/java/org/hibernate/type/TypeFactory.java]: I could also investigate whether I could use a bootstrap finished callback to clear these registers and free up memory. Do you think it's worth researching it a little bit?

**Comment by vladmihalcea** [comment]
@sebersole I pushed a new version which allows us to clear the registries via the `SessionFactoryObserver`. Let me know what you think.

**Pull request merged**